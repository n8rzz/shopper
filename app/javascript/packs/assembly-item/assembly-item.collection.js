export default class AssemblyItemCollection {
    get length() {
        return this._items.length;
    }

    constructor() {
        this._items = [];

        return this._init();
    }

    _init() {
        return this;
    }

    add(viewModel) {
        this._items.push(viewModel);
    }

    findById(itemId) {
        return this._items.filter((item) => item.itemId === itemId)[0];
    }
}
