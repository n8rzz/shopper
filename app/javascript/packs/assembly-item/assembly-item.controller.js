import ApiService from '../service/api.service';
import AssemblyItemCollection from './assembly-item.collection';
import AssemblyItemView from './assembly-item.view';

const CLASSNAMES = {
    TRIGGER_ELEMENT: 'js-addAssemblyItemToOrder',
};

export default class AssemblyItemController {
    constructor() {
        const $elementList = document.getElementsByClassName(CLASSNAMES.TRIGGER_ELEMENT);
        this._collection = new AssemblyItemCollection();
        this._isEnabled = false;

        if ($elementList.length === 0) {
            return this.disable();
        }

        return this._init()
            ._setupHandlers()
            ._createChildren($elementList)
            .enable();
    }

    _init() {
        return this;
    }

    _setupHandlers() {
        this._onClickHandler = this._onClick.bind(this);

        return this;
    }

    _createChildren($elementList) {
        for (let i = 0; i < $elementList.length; i++) {
            const element = $elementList[i];
            const viewModel = new AssemblyItemView(element, this._onClickHandler);

            this._collection.add(viewModel);
        }

        return this;
    }

    enable() {
        if (this._isEnabled) {
            return this;
        }

        this._isEnabled = true;

        return this;
    }

    disable() {
        if (!this._isEnabled) {
            return this;
        }

        this._isEnabled = false;

        return this;
    }

    _onClick(event) {
        event.preventDefault();

        const csrf = event.currentTarget.dataset.csrf;
        const assemblyId = event.currentTarget.dataset.assemblyId;
        const assemblyItemId = event.currentTarget.dataset.assemblyItemId;
        const departmentId = event.currentTarget.dataset.departmentId
        const assemblyItemUrl = `/order_items/create/item.json`;
        const requestPayload = {
            assembly_id: assemblyId,
            department_id: departmentId,
            item_id: assemblyItemId
        };

        ApiService.post(assemblyItemUrl, requestPayload, csrf)
            .then((response) => {
                if (response.status >= 300) {
                    console.error(`Received an unexpected status code from ${assemblyItemUrl}. See response: `, response);

                    return;
                }

                this._onAddItemSuccess(requestPayload, response);
            })
            .catch((error) => { throw error; });
    }

    _onAddItemSuccess(requestPayload, response) {
        console.log(requestPayload, response);
    }
}

document.addEventListener('turbolinks:load', () => {
    const assemblyItemController = new AssemblyItemController();
});
