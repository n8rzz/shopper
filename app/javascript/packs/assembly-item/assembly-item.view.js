export default class AssemblyItemView {
    get assemblyId() {
        return this._assemblyId;
    }

    get itemId() {
        return this._itemId;
    }

    constructor($element, onClickHandler) {
        this._$element = $element;
        this._assemblyId = $element.dataset.assemblyId;
        this._itemId = $element.dataset.assemblyItemId;
        this._isEnabled = false;

        if (this._$element == null) {
            return this.disable();
        }

        return this._init()
            ._createChildren()
            ._setupHandlers(onClickHandler)
            .enable();
    }

    _init() {
        return this;
    }

    _createChildren() {
        return this;
    }

    _setupHandlers(onClickHandler) {
        this._onClickHandler = onClickHandler;

        return this;
    }

    enable() {
        if (this._isEnabled) {
            return;
        }

        this._$element.addEventListener('click', this._onClickHandler);

        this._isEnabled = true;

        return this;
    }

    disable() {
        if (!this._isEnabled) {
            return;
        }

        this._$element.removeEventListener('click', this._onClickHandler);

        this._isEnabled = false;

        return this;
    }
}
