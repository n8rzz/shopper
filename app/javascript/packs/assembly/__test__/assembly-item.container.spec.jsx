import React from 'react';
import { shallow } from 'enzyme';

import { AssemblyItemContainer } from '../assembly-item.container';
import {
    assemblyMock,
    assemblyItemsGroupedByAssemblyIdMock,
    csrfMock,
    departmentMapMock,
    itemMapMock,
} from '../../__mocks__/mocks';

describe('AssemblyItemContainer', () => {
    test('renders correctly with valid props', () => {
        const component = shallow(<AssemblyItemContainer
            assembly={assemblyMock}
            assemblyItemsGroupedByAssemblyId={assemblyItemsGroupedByAssemblyIdMock}
            csrf={csrfMock}
            departmentIdMap={departmentMapMock}
            itemIdMap={itemMapMock}
        />);

        expect(component).toMatchSnapshot();
    });
});
