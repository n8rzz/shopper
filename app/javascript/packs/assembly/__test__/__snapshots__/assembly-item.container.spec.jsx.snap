// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssemblyItemContainer renders correctly with valid props 1`] = `
<AssemblyItem
  assembly={
    Object {
      "created_at": "2019-07-29T22:49:47.431Z",
      "id": 1,
      "name": "Souvlaki",
      "updated_at": "2019-07-29T22:49:47.431Z",
    }
  }
  assemblyItemsGroupedByAssemblyId={
    Object {
      "1": Array [
        Object {
          "assembly_id": 1,
          "created_at": "2019-08-01T22:58:09.157Z",
          "id": 1,
          "item_id": 13,
          "qty": 8,
          "updated_at": "2019-08-01T22:58:09.157Z",
        },
      ],
    }
  }
  departmentIdMap={
    Object {
      "3": Object {
        "created_at": "2019-07-21T16:03:32.883Z",
        "id": 3,
        "name": "Bakery",
        "updated_at": "2019-07-21T16:03:32.883Z",
      },
      "5": Object {
        "created_at": "2019-07-21T16:03:32.883Z",
        "id": 5,
        "name": "Meat",
        "updated_at": "2019-07-21T16:03:32.883Z",
      },
    }
  }
  isAssemblyItemSubmitSuccess={false}
  isOpen={false}
  isSubmitAssemblySuccess={false}
  isSubmittingAssembly={false}
  itemIdMap={
    Object {
      "13": Object {
        "created_at": "2019-07-21T16:03:33.162Z",
        "department_id": 3,
        "id": 13,
        "name": "Agave Syrup",
        "updated_at": "2019-07-21T16:03:33.162Z",
      },
      "15": Object {
        "created_at": "2019-07-21T16:03:33.162Z",
        "department_id": 5,
        "id": 15,
        "name": "Agave Syrup",
        "updated_at": "2019-07-21T16:03:33.162Z",
      },
    }
  }
  onClickAccordionTriggerHandler={[Function]}
  onClickAddAssemblyHandler={[Function]}
  onClickAddAssemblyItemHandler={[Function]}
  submittedAssemblyItemId={-1}
/>
`;
